### ISSUE TITLE
Papercut refresh glitches

### ISSUE THREAD

**Issue by poire-z** [koreader/koreader#3130]
* KOReader version: latest nightly, but this has been happening since I discovered KOReader
* Device: Kobo Glo HD

Every now and then, i get this kind of display glitch, some white triangle zone at top left of screen :

![glohd_papercut](https://user-images.githubusercontent.com/24273478/29665498-9535e59a-88d4-11e7-86a1-2a2f5a976932.jpeg)

(Nothing dramatic, as a diagonal swipe to trigger a full refresh make it dissapear.)
It mostly happens when going into TOC (two finger swipes right), or closing it.
It also happens when changing fonts (where there's a quite a few parallele updates happening : the drawing of the page with the new font, the update of the menu with checkboxes adjusted, the "Redrawing with <font name>" notification, and its disapearance after a 1 second timeout).

This kind of problem has been mentionned in #1359 (and probably other issues) as some refreshes collisions, and I followed @hwhw hints and changed:
https://github.com/koreader/koreader-base/blob/e1974cbc8841dbdff2f837da1d2cde6ec6b107e3/ffi/framebuffer_mxcfb.lua#L414-L417
to have them all set at `true` (and added `self.wait_for_marker_ui = true` there too.
Since I did this, I haven't witness such glitch (changing only a few of them to true, but not all, did not make it not happen).
I don't feel any less smoothness, but **I wonder if there may be some side effects with having all refreshes being waited for (battery usage?)**

But I'm not sure this is the right fix for it.
It may be that the TOC and setFont codes may not be using the right `setDirty()` and such, as I only have this problem with these.
**Before digging much on this side, I'd like to know:**
**- do other people with Glo HD witness this glitch too, and when doing what actions ?**
**- do people with other Kobo or Kindle devices witness this too ?**

**Comment by KenMaltby** [comment]
That would scare the heck out of me, it resembles what you can see with a cracked substrata.  I am also still surprised that the Glo HD can do two finger gestures.  I have not noticed any such issue on my AuraHDs.

**Comment by Frenzie** [comment]
I have very occasionally seen an odd effect similar to that on my H2O. The screen might be physically different than yours as I think it occurs in the top-right corner in the very rare, ahem, edge case.

**Comment by cramoisi** [comment]
@poire-z  Got this each time I manually suspend (aka pushing the power button) my readers for displaying a random image since my first install of koreader.

**Comment by poire-z** [comment]
Are these glitches transient (you see them for 0.5 second) or permanent (you need a full refresh to make them disappear)?
I sometimes get a transient one when _resuming_ from screensaver.
I think I managed to not get these anymore while changing fonts by just commenting these lines (in my own patch):
https://github.com/koreader/koreader/blob/e8aab49ee91dda8830de8023066d5cdd3c63ccc5/frontend/apps/reader/modules/readerfont.lua#L248-L251
This notification + the redraw of the menu for checking the box of the new fonts + the redraw of the page below were, I think, making refreshes traffic very busy :) and causing this glitch.
I still wonder if it's a GloHD thing or if other devices get these too.

**Comment by cramoisi** [comment]
@poire-z always transient : it happens each time before displaying a full screen image (like when sleeping, rebooting and poweroff)

**Comment by poire-z** [comment]
Oh, yes, I had seen that FileManager effect too: when changing fonts, I sometimes got a slice of the filemanager transiently displayed on the font name I taped :)
I think it's because, when going from FileManager to Reader, the FileManager is not closed, it's still a widget in the widget stack, it's just really below the others (the reader is obviously above, taking the full screen, hiding it). But when refreshing, each widget in the full widgets stack is asked to paint itself... So, if some `scheduleIn()` or `nextTick()` are used and UIManager gets to draw some of them: even more busy refresh traffic.

**Comment by cramoisi** [comment]
Yup - it seems random. There is something jinxed about it : when I don't care, it happen each time and when I try to narrow-testing it, it doesn't happen. Got that too with the landscape / portrait FM.

**Comment by poire-z** [comment]
@cramoisi : do you still get these glitches after #3770 ? (not sure it did solve them, just curious in case it did)

**Comment by Frenzie** [comment]
I just saw one yesterday I iirc.

**Comment by cramoisi** [comment]
@poire-z I far less than before (now very few when going to sleep / always before) but I saw them at other moments (when changing font, etc)

**Comment by poire-z** [comment]
Following up on https://github.com/koreader/koreader/issues/3833#issuecomment-379461278 here:

Strange that you get this on the top right side :) It's always on the top left side of my GloHD when it happens.
May be there's some mirror x/y or coordinates translation differences between your H2O and the GloHD.

Anyway, I still get it, but not always, when changing font.
It's probably some concurrent refreshes (like if refreshes are like tide coming from some corner, which is soon caught up and aborted by the next tide, which could explain the triangle).

I tried in the past to change some framebuffer settings by adding this into `frontend/device/kobo/device.lua `:
```lua
 function Kobo:init()
     self.screen = require("ffi/framebuffer_mxcfb"):new{device = self, debug = logger.dbg}
+    -- hijack the defaults set in ffi/framebuffer_mxcfb.lua:
+    if G_reader_settings:isTrue("eink_wait_for_all_markers") then
+        self.screen.wait_for_marker_partial = true
+        self.screen.wait_for_marker_fast = true
+        self.screen.wait_for_marker_ui = true
+    end
     self.powerd = require("device/kobo/powerd"):new{device = self}
```
but it didn't solve these papercut glitches.

About this glitch on suspend you and @cramoisi are getting and that I don't, I see that I have this single line commented in my own koreader mods:
https://github.com/koreader/koreader/blob/7a28c7d7aa5a38e6f3acb5ee9246017c96e8d8ff/frontend/device/generic/device.lua#L141
with an old comment saying  "this refreshFull is not necessary with our full screen screensaver images" - dunno if my reasoning at the time still applies with the recent screensaver rework - and if there would be side effects with other screensaver modes if it were to be commented out.
You may want to check if commenting it solves these glitches on your devices.

Note that the previous line:
https://github.com/koreader/koreader/blob/7a28c7d7aa5a38e6f3acb5ee9246017c96e8d8ff/frontend/device/generic/device.lua#L140
if there is a screensaver widget to show, would generate at some point a full refresh too:
https://github.com/koreader/koreader/blob/7a28c7d7aa5a38e6f3acb5ee9246017c96e8d8ff/frontend/ui/screensaver.lua#L284

**Comment by cramoisi** [comment]
@poire-z top left side, always.

**Comment by poire-z** [comment]
> top left side, always.

@cramoisi: was that a general comment, or do you mean you still get it after commenting the line I suggested:
https://github.com/koreader/koreader/blob/7a28c7d7aa5a38e6f3acb5ee9246017c96e8d8ff/frontend/device/generic/device.lua#L141

@Frenzie : I don't get any glitch when going to screensaver with your djvu :|
Could you check if you still get it after commenting the line I suggested?

**Comment by cramoisi** [comment]
@poire-z : that was a general comment for the gloHD and KA1.

**Comment by NiLuJe** [comment]
I vaguely remember hitting a few of those when looking into the waveform update modes on the Kobo, so I'm not entirely convinced the blame is entirely on our side, but it is definitely related to concurrent refreshes.

**Comment by cramoisi** [comment]
@poire-z I'm not really see it anymore even without commenting the line ; only in very specific cases (like when testing new hyphenation patterns) and even then it's very fast. I will comment the line now and give you a feedback this night.

**Comment by cramoisi** [comment]
@poire-z @Frenzie  : I didn't see the glitch today...

**Comment by cramoisi** [comment]
@poire-z got __one__ papercut glitch yesterday when over-switching hyph patterns

**Comment by NiLuJe** [comment]
Which FW are you running? (Not that the kernels actually ever get updated, but you never know...).

I just did a quick check on my H2O FW 4.7 (`Linux (none) 2.6.35.3-850-gbc67621+ #3032 PREEMPT Mon Jan 9 13:37:40 CST 2017 armv7l GNU/Linux`), nickel still doesn't MXCFB_WAIT_FOR_UPDATE_COMPLETE for UI elements, but does it for UPDATE_MODE_PARTIAL.

One thing it does for the selection highlight in a menu is do it with the B&W waveform (NTX_WFM_MODE_A2), which hopefully we're also doing with waveform_fast ;).

Anyway, see if setting `self.wait_for_marker_partial` to `true` for your Kobo model in `ffi/framebuffer_mxcfb.lua` help?

**Comment by cramoisi** [comment]
@NiLuJe on my gloHD, 3.16 and it's happen very often, on my KA1, 4.4.9298 I think (before the sdcv problem, I had planned to update the gloHD to the latest

**Comment by cramoisi** [comment]
> Anyway, see if setting self.wait_for_marker_partial to true for your Kobo model in ffi/framebuffer_mxcfb.lua help?

There is no kobo_daylight model, should I add one ?

**Comment by NiLuJe** [comment]
@cramoisi : Yep, you'll have to add the full model-specific block. Note that, thankfully, the defaults match 95% of the Kobo devices.

Still, I'd like to fix that, is there a list of the various Kobo model nicknames somewhere?

**Comment by cramoisi** [comment]
@NiLuJe if there is one, tshering@mobileread must know where. I usually open the KSM launch script to find them :)

oh, and : https://www.mobileread.com/forums/showpost.php?p=3472982&postcount=49

(the h2O2 is missing I think)

**Comment by NiLuJe** [comment]
@cramoisi Thanks, I'll take a look at KSM's script then ;).

**Comment by Frenzie** [comment]
Pssht, we also have a list:

https://github.com/koreader/koreader/blob/9e4e71d63016d520eab1d59cc07e644958c63afe/frontend/device/kobo/device.lua#L544-L565

(The comments explain which models are which.)

**Comment by NiLuJe** [comment]
Netx step: blindly guess if any of these still use the mangled REAGL support of the original Aura? ^^

**Comment by Frenzie** [comment]
lol, trying to push me into tagging a release as stable with upcoming huge breaking changes, are we :P

**Comment by cramoisi** [comment]
so : 

`elseif self.device.model == "Kobo_daylight" then
            self.wait_for_marker_full = true
            self.wait_for_marker_partial = true
            self.wait_for_marker_fast = false`

**Comment by NiLuJe** [comment]
@cramoisi : See 146be8c4fecaa71b80ce8377215a9f3dc91e8113

@Frenzie : :P. Nah, I wasn't feeling that adventurous, just flagged a couple of FIXMEs ;p.

**Comment by poire-z** [comment]
I added this to my own patch, to quickly test those:
```lua
--- orig_koreader/frontend/device/kobo/device.lua       2018-04-14 08:01:22.000000000 +0200
+++ koreader/frontend/device/kobo/device.lua    2018-04-14 09:18:40.000000000 +0200
@@ -148,6 +148,29 @@

 function Kobo:init()
     self.screen = require("ffi/framebuffer_mxcfb"):new{device = self, debug = logger.dbg}
+    -- pour éviter "papercut refresh glitches"
+    -- (on hijack ce qui est dans ffi/framebuffer_mxcfb.lua)
+    if G_reader_settings:isTrue("eink_wait_for_all_markers") then
+        self.screen.wait_for_marker_partial = true
+        self.screen.wait_for_marker_fast = true
+        self.screen.wait_for_marker_ui = true
+    end
+    -- pour voir si ce mode (utilisé sur kindle) est bien
+    -- (on hijack ce qui est dans ffi/framebuffer_mxcfb.lua)
+    if G_reader_settings:isTrue("eink_enable_reagl") then
+        local ffi = require("ffi")
+        self.screen.waveform_partial = ffi.C.WAVEFORM_MODE_REAGL
+        self.screen.update_mode_partial = ffi.C.UPDATE_MODE_FULL -- REAGL get upgraded to full
+        self.screen.wait_for_marker_partial = true
+        self.screen.wait_for_marker_fast = true
+    end
+    if G_reader_settings:isTrue("eink_enable_reagld") then
+        local ffi = require("ffi")
+        self.screen.waveform_partial = ffi.C.WAVEFORM_MODE_REAGLD
+        -- self.screen.update_mode_partial = ffi.C.UPDATE_MODE_FULL -- no need for REAGLD, sinon flash
+        self.screen.wait_for_marker_partial = true
+        self.screen.wait_for_marker_fast = true
+    end
     self.powerd = require("device/kobo/powerd"):new{device = self}
     self.input = require("device/input"):new{
         device = self,
```
On my GloHD FW 3.16, the first didn't change the papercut glitch appearance.

Enabling REAGLD does not change anything.

But I was surprised REAGL worked: it's kind of different when changing pages, letters morphing into the next ones, may be quicker. But I didn't like it: when changing pages, the footer does not stay unrefreshed, it's fully going blank then being redrawn quickly. I wonder if it's the same on Kindle.
Note that we also changed a bunch on `setDirty(..., "partial")` into `setDirty(..., "ui")` to avoid using REAGL for some stuff on Kindle (full story in #3230).

**Comment by NiLuJe** [comment]
@poire-z : IIRC, the behavior you're describing is what happens on devices which don't actually really handle REAGL. Take that with a grain of salt, it's been a veeeery long time since I tried this ^^.

**Comment by cramoisi** [comment]
erf, got a crash when waking up, will open a issue with the log

**Comment by NiLuJe** [comment]
Wait, no, those fugly-ass ifdefs got me. They didn't break ABI, they just added a *new* ioctl on !Android.
Yay.

**Comment by cramoisi** [comment]
@NiLuJe when setting https://github.com/NiLuJe/koreader-base/blob/146be8c4fecaa71b80ce8377215a9f3dc91e8113/ffi/framebuffer_mxcfb.lua#L430 to true, turning pages is absurdly slow ...

**Comment by NiLuJe** [comment]
Hmm, either it's expecting the new wait_for ioctl, or... something, I guess?

More on that once I have a working strace patch (in.... a while).

**Comment by cramoisi** [comment]
@poire-z when i use your patch and enable the eink_enable_reagl, when I turn page, it goes forward and backward without reason (when it should goes forward)

**Comment by poire-z** [comment]
Never experienced that, and it shouldn't, this patch can't generate any page forward/backward event/request...
On which device, HD or One? On my GloHD, it has this strange effect:
> But I was surprised REAGL worked: it's kind of different when changing pages, letters morphing into the next ones, may be quicker. But I didn't like it...

May be it's that morphing that makes you feel that? :|
If not, that' super strange. May be the device re-display the previous screen buffer... I would try to tap forward 4 times (so, you should still see the same initial page with what you describe), then exit koreader, and launch it again, and see if it starts on the same initial page, or if you are 4 page further.

**Comment by cramoisi** [comment]
Also, I updated my devices (gloHD and KA1, I won't touch to my wife's aura_v1 3.12.1) to the latest firmware (just to see, but there is nothing to see only they updated their french dict to petit robert 2013, (but perhaps it was before as I don't use nickel) which is nice.

**Comment by NiLuJe** [comment]
I *finally* managed to finish updating my strace patch, see [BASE#390](https://github.com/koreader/koreader-base/issues/390#issuecomment-388592300).

Now I need guinea pigs with Mark6 & Mark7 devices to check if we're doing things right ;).

**Comment by cramoisi** [comment]
@NiLuJe : got a Mark6 to test but didn't have ssh access from nickel; any directions about that ?

**Comment by cramoisi** [comment]
@poire-z ; running the script in koreader or ksm before launching nickel don't work.

**Comment by NiLuJe** [comment]
@cramoisi : Make sur you don't forget to background it (`&`).

If you're fond of Very Big Hammers, there's always my [usual stash of crap](https://www.mobileread.com/forums/showthread.php?t=254214) that includes a telnet & ssh server (which will persists through FW updates).

[Here](http://files.ak-team.com/niluje/mrpub/Kobo/kobo-stuff-1.4.N-r14885.tar.xz) is an updated package, which includes said strace changes, among a whole host of other things.

As usual in Kobo-land, uninstall can get a bit tricky, but you should be able to "unplug" everything from the boot process simply by deleting `/etc/udev/rules.d/99-stuff.rules`.

**Comment by onde2rock** [comment]
> EDIT²: Shot in the dark: there's also the recent addition of the possibility of invoking a custom dropbear instance from KOReader itself. Does it persist after KOReader's exit (ping @onde2rock)?

Yes, it's running it's own daemon, it doesn't depend on KOReader.

**Comment by NiLuJe** [comment]
Great! Then that's probably the least intrusive solution for people not already running SSH stuff permanently ;).

**Comment by cramoisi** [comment]
Could somebody tells me where is the & on KOReader terminal's emulator ?

**Comment by poire-z** [comment]
Running this in the terminal emulator works for me:
`(/mnt/onboard/.adds/misc/bg_strace.sh >/dev/null 2>&1 &)`

**Comment by cramoisi** [comment]
@NiLuJe 
[trace.log](https://github.com/koreader/koreader/files/2006338/trace.log)

**Comment by cramoisi** [comment]
@poire-z yup by launching koreader with ksm > koreader additions > start koreader nickel switch. Then in koreader, when you burger > Exit > Exit , it automatically launch nickel

**Comment by NiLuJe** [comment]
@cramoisi : Thanks, that's exactly the device I was looking for (since it's arguably the most uncommon in terms of screen).

Good news: it appears to be behaving like Mark5's too, so, we're probably good :).

**Comment by cramoisi** [comment]
@poire-z no need to do the gloHD as @poire-z did it. but I got an old kobo aura (vers 1). Interested ?

**Comment by poire-z** [comment]
@cramoisi : you could do your GloHD if it has a recent firmware (I did it for a very old 3.16). May be kobo do stuff differently in recent firmwares (?)

**Comment by cramoisi** [comment]
ok for the gloHD (last pulled firmware : 4.8.10956 ) ! the aura v1 has a very old firmware though (3.12.2 iirc)

**Comment by NiLuJe** [comment]
I doubt the Glo HD will be any different from @poire-z 's or your KA1, but if it's not too much bother, it's FOR SCIENCE!

The Aura might be very interesting, too, since I can't recall if I ever got someone to do that, or if I winged that bit... (Because, supposedly, that one *does* things slightly differently, hence the different 'Full flash after N page turns' menu).

**Comment by cramoisi** [comment]
no problem, I'll do it tomorrow for I need to get my hands on my wife's reader and it won't be easy but it's impossible after 8PM ;-)

**Comment by cramoisi** [comment]
@NiLuJe Kobo Aura v1, FW 3.12.1
[trace.log](https://github.com/koreader/koreader/files/2010266/trace.log)

**Comment by NiLuJe** [comment]
Yup, that one is using REAGL, as expected.

And I don't think we're doing it the exact same way, so I probably never had such a log, yay \o/.

**Comment by NiLuJe** [comment]
@cramoisi : If you can manage to wrestle that device from your better half again, could you check if [BASE#676](https://github.com/koreader/koreader-base/pull/676) behaves better?

**Comment by cramoisi** [comment]
@NiLuJe not before tomorrow I'm afraid. But I think I could manage to do it before diner.

> I'm going to file the last four EINVALID returns as nickel dying and not something going terribly wrong with my strace patch ^^.

Indeed, I missed the exit nickel png and touch the shutdown device one

**Comment by NiLuJe** [comment]
Also, in a more general manner, we're barely doing any real UPDATE_PARTIAL. Most of our partials target the full viewport, either because our menus are of variable size, so it makes using the proper region size a tricky proposition, or because the small UI refreshes get folded into a menu swap.

**Comment by cramoisi** [comment]
@NiLuJe : can I clone and compile a kobo koreader directly from your https://github.com/NiLuJe/koreader/tree/uimanager-wip ?

**Comment by NiLuJe** [comment]
@cramoisi : It'll only ever affect the Lua bits, so you're probably better off just building a patch out of the 'new' commits and applying that on top of whatever your want. (I won't update submodules, so you'd end up with outdated stuff if you were to build from the branch directly).

**Comment by cramoisi** [comment]
@NiLuJe cool, because I compile a kobo release yest and it compiled right but crashes on the device.
It's only the change in framebuffer_mxcfb.lua ?

**Comment by NiLuJe** [comment]
@cramoisi : `framebuffer_mxcfb.lua` is from base, so, no? That'd be PR [BASE#676](https://github.com/koreader/koreader-base/pull/676).

**Comment by NiLuJe** [comment]
I worked a bit on said branch today, so it's been rebased. It's also slightly less broken, but now *depends* on the PR I just mentioned.

It's basically a giant game of whack-a-mole, but hopefully I'll get there...

**Comment by cramoisi** [comment]
I'll put it on the aura today (i'll take it with me at the office), any thing to look at ?

**Comment by NiLuJe** [comment]
Mainly the reader, since that's where reagl is used the most ;).

The rest of the branch is still a wip, and between the pw2 and the H2O, I
can cover the two extremes ;). So mainly making sure the base PR didn't
break anything on the Aura :).

**Comment by cramoisi** [comment]
Ok , i'll put it on today, will do a global check at lunch and will give it back to my wife (and wait to see if it breaks her evening reading session :-)

**Comment by NiLuJe** [comment]
Okay, after #3983, the only place I can reliably trigger this kind of thing is with the virtual keyboard:

* Generally spamming keys can lead to them being "blanked", but that's an inherent limitation of using monochrome update modes.
* But, by holding the "return" key to delete a string of input, I can somewhat reliably trigger a diagonal tear in the top-left corner of the keyboard.
I think I even reproduced it on a Kindle ;o).

**Comment by Frenzie** [comment]
Kindle? Cool :P

**Comment by cramoisi** [comment]
saw one or two cut for microseconds when enabling floating punctation or changing hyphenation but really nothing compared to before (and I didn't comment back the l[ine 141 ](https://github.com/koreader/koreader/blob/7a28c7d7aa5a38e6f3acb5ee9246017c96e8d8ff/frontend/device/generic/device.lua#L141) after updating to 1735) and, above all, nothing which stays =)

**Comment by Biep** [comment]
I have this on my Kobo Aura One, and the triangle is permanent - i.e. until I refresh one way or another.
I have seen lots of them in a short space when browsing a large PDF (i.e. tapping the top left corner and then sliding down).
KOReader version: v2019-02-130-g93422d0-2019-03-02

Glad to see it is not some hardware fault.

**Comment by NiLuJe** [comment]
That's.. surprising. I wouldn't mind a picture, and a *very* specific chain of events than can trigger it.

**Comment by Biep** [comment]
I should get around to trying to create the effect again - when I have a 'phone with a camera so that I can take a picture.

**Comment by Biep** [comment]
I just got a new type, though. While deleting letters from the dictionary lookup feature (popping up when long pressing an entry head). When I had deleted all letters and then hit the <x button once more, a triangular slice of the screen went blank. The two top corners were 1/4 and 1/2 along the top rim, and the bottom corner was the SW corner of the screen.

**Comment by NiLuJe** [comment]
Yeah, the virtual keyboard can still sometimes trigger weird cases like that, that one is not terribly surprising ;).

**Comment by Biep** [comment]
No, but the _shape_ was.  Not just a "dog-ear".

**Comment by gerhaher** [comment]
Another variant: trapezoids. 
Happens now and then for pdf's, especially when I perform full text searches.

Looks something like this:

![Reader_2019-Aug-28_221229 (1)](https://user-images.githubusercontent.com/32645430/64185752-f8a79480-ce6d-11e9-8ecc-0b4003b98cb7.png)

**Comment by NiLuJe** [comment]
@gerhaher : On what kind of device & FW, for reference?

**Comment by gerhaher** [comment]
Device: KA1
FW: v2019.08-88-gb2554ba_2019-09-01/, (or 4.15.12920 if you mean Nickel)

**Comment by Frenzie** [comment]
Here's one I've never seen before.

![P_20201222_220059](https://user-images.githubusercontent.com/202757/102932872-721f2800-44a1-11eb-8ed2-be6fd4b18e29.jpg)

**Comment by NiLuJe** [comment]
Yeah, that might happen more than before with stacked IMs since https://github.com/koreader/koreader/pull/6828

**Comment by jospalau** [comment]
I hadn't noticed this problem in 2 years but lately it has happened a few times both in the Libra 2 and Clara 2E.

I only can reproduce it when I start using the device and it looks it happens after applying a profile with multiple options with stacked messages.

![IMG_20240613_091343](https://github.com/koreader/koreader/assets/85340/a5844ea8-9241-43d7-9f04-5667bce72fe0)
